### PR TITLE
Prevent potential buffer hang

### PR DESCRIPTION
**Pull request by JasonLG1979**
React to the buffer message immediately and also fire a short repeating timeout to check the buffering state that cancels only if we're not buffering or there's a pending timeout. This will insure we don't get stuck in a buffering state if we're really not buffering.

### CODE REVIEW

File: pithos/pithos.py [hunk 1]
```diff
@@ -861,12 +862,29 @@ def check_if_song_is_ad(self):
...
+        # React to the buffer message immediately and also fire a short repeating timeout
+        # to check the buffering state that cancels only if we're not buffering or there's a pending timeout.
+        # This will insure we don't get stuck in a buffering state if we're really not buffering.
+
+        self.react_to_buffering_mesage(False)
+
+        if self.buffering_timer_id:
+            GLib.source_remove(self.buffering_timer_id)
+            self.buffering_timer_id = 0
+        self.buffering_timer_id = GLib.timeout_add(200, self.react_to_buffering_mesage, True)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/pithos.py [hunk 2]
```diff
@@ -861,12 +862,29 @@ def check_if_song_is_ad(self):
...
+    def react_to_buffering_mesage(self, from_timeout):
         # If the pipeline signals that it is buffering set the player to PseudoGst.BUFFERING
         # (which is an alias to Gst.State.PAUSED). During buffering if the user goes to Pause
         # or Play(an/or back again) go though all the motions but don't actaully change the
         # player's state to the desired state until buffering has completed. The player only
         # cares about the actual state, the rest of Pithos only cares about our buffer_recovery
         # state, the state we *wish* we were in.
+
+        # Reset the timer_id only if called from the timeout
+        # to avoid GLib.source_remove warnings.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [W291] trailing whitespace

**Comment by JasonLG1979** [comment]
OK @TingPing It's telling me that something has failed but not what? What do I need to fix?

File: pithos/pithos.py [hunk 3]
```diff
@@ -861,12 +862,29 @@ def check_if_song_is_ad(self):
...
+        # React to the buffer message immediately and also fire a short repeating timeout
+        # to check the buffering state that cancels only if we're not buffering or there's a pending timeout.
+        # This will insure we don't get stuck in a buffering state if we're really not buffering.
+
+        self.react_to_buffering_mesage(False)
+
+        if self.buffering_timer_id:
+            GLib.source_remove(self.buffering_timer_id)
+            self.buffering_timer_id = 0
+        self.buffering_timer_id = GLib.timeout_add(200, self.react_to_buffering_mesage, True)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [W291] trailing whitespace

Comment (JasonLG1979) [inline pithos/pithos.py]: Finally... ;P

**Comment by levibostian** [comment]
I pulled down your pithos repo, @JasonLG1979. I checked out this branch, merged master, and am runnning that Pithos instance now. Giving it a try to see if it indeed solves the buffering issue. I'll report back if I have issues today.

**Comment by JasonLG1979** [comment]
Right on. I'm pretty sure it fixes the issue. I really wish Gstreamer network buffering didn't suck...

**Comment by levibostian** [comment]
Music just froze playback. After about an hour and a half playback time non-stop. 

Pithos does not show "Buffering..." as this issue is trying to fix. Instead, the song playback stopped at this moment of the song:

img_20170203_103129

It has been stuck here about 10 minutes. I double click the song below it and it starts playing again fine. I'll keep track of this more though the day. 

The console log looks clean even after the stop in playback:

img_20170203_103144

**Comment by JasonLG1979** [comment]
That looks a lot like issues we were dealing with a couple years ago. It's a Gstreamer issue. If you're running Raspbian you have a pretty old version of Gstreamer that still has that issue.

**Comment by levibostian** [comment]
I am doing a `dist-upgrade` now as we speak.

**Comment by JasonLG1979** [comment]
> I am doing a dist-upgrade now as we speak.

You can dist-upgrade all you want. You'll never get a current version of Gstreamer in Debian stable.

**Comment by levibostian** [comment]
> You can dist-upgrade all you want. You'll never get a current version of Gstreamer in Debian stable.

Well, that sucks.

**Comment by JasonLG1979** [comment]
> Well, that sucks.

It's by design. That's why it's called stable. They freeze the versions of everything and only backport bug fixes(if you're lucky).

**Comment by JasonLG1979** [comment]
Debian's great IMHO for a server but as a desktop OS it get's stale real quick. Even a brand new release will have packages at least one version old. Debian Jessie was released in April 2015 so you're more than likely dealing with package versions from some time in 2014.

**Comment by levibostian** [comment]
That's understandable. I appreciate stable software. 

Looks like I don't have a solution to this unless I want to run on something other than Raspbian. From the dev team's experience, what is the minimum GStreamer version recommended for Pithos?

I will keep you updated on "Buffering..." if that happens again.

**Comment by JasonLG1979** [comment]
I can't speak for everyone because I'm not a part of the "dev team" I'm just a contributor(really pretty much the only one,lol!!!) the "dev team" would be @TingPing. But from my experience 1.6 seemed to work fine and from 1.8 on I've had very little problems. I have yet to have a version that worked perfectly.

**Pull request merged**